lib/heap.ex

```diff
@@ -196,6 +230,13 @@ defmodule Heap do
   defp meld({k0, l0}, {k1, _} = r, :<) when k0 > k1, do: {k0, [r | l0]}
   defp meld({_, _} = l, {k1, r0}, :<), do: {k1, [l | r0]}
 
+  defp meld({k0, l0} = l, {k1, r0} = r, fun) do
+    case fun.(k0, k1) do
+      true -> {k0, [r | l0]}
+      false -> {k1, [l | r0]}
+    end
```

Comment:
should there be a more friendly error than function clause error if the function returns a non-boolean value?

Reply:
I'm busy looking at how `Enum.sort/2` handles bad sort functions, it matches a sub function based on `true` or `false` parameters. I could do the same?

```elixir
iex(1)> Enum.sort 1..3, &(&1 + &2)
** (FunctionClauseError) no function clause matching in Enum.sort_merge/3    
    
    The following arguments were given to Enum.sort_merge/3:
    
        # 1
        [[2, 1, 3]]
    
        # 2
        &:erlang.+/2
    
        # 3
        3
    
    Attempted function clauses (showing 2 out of 2):
    
        defp sort_merge(list, fun, true)
        defp sort_merge(list, fun, false)
    
    (elixir) lib/enum.ex:3005: Enum.sort_merge/3
```

I could also raise a RuntimeError?

```
    case fun.(k0, k1) do
      true -> {k0, [r | l0]}
      false -> {k1, [l | r0]}
      err -> raise("Comparator should return boolean, but returned '#{err}'.") 
    end
```

Reply:
I think raising a runtime error is fine.